Hold a mutex while refreshing the token

flub opened the pull request:
When using the default service account from the internal metadata
service there is no point in asking it for multiple tokens at the same
time.  With a high enough request rate this results in a thundering
herd of refresh requests and eventually the metadata service rejecting
the requests with HTTP 429 status code.

Commit:
Hold a mutex while refreshing the token

When using the default service account from the internal metadata
service there is no point in asking it for multiple tokens at the same
time.  With a high enough request rate this results in a thundering
herd of refresh requests and eventually the metadata service rejecting
the requests with HTTP 429 status code.

loewenheim commented on src/default_service_account.rs:
```diff
@@ -13,6 +13,10 @@ use crate::util::HyperExt;
 pub(crate) struct DefaultServiceAccount {
     token: RwLock<Token>,
+    // In theory we should be able to use a single RwLock for this, but when refreshing we
+    // go accross await points.  Always using a tokio RwLock is inefficient in the hot path
```
```suggestion
    // go across await points.  Always using a tokio RwLock is inefficient in the hot path
```

Swatinem reviewed (approved):
I guess this works, but is not particularly nice -_-

flub commented:
> I guess this works, but is not particularly nice -_-

i agree, i was hoping someone else would have a better idea :)

Commit:
Update src/default_service_account.rs

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

Commit:
Add same for authorised user

Commit:
Move refersh locking to AuthenticationManager

This makes it central in once place.  The downside is that it is a
global lock also for the CustomServiceAccount case which could
theorethically have a lock per scope combination requested.

Smarter approaches are possible, but this is a good first approach to
avoid a thundering herd on the refresh.

flub commented:
PTAL, i've changed this to be on a different level.  otherwise i think this is good enough for us.
